### PR TITLE
feat: Support multiple table functions in queries

### LINKED ISSUE #3513: feat: Extend table function implementation to support multiple functions in SELECT
Parent issue: https://github.com/confluentinc/ksql/issues/3504

Extend the implementation to support multiple table functions in the SELECT clause.

### PR DESCRIPTION
**Pull request by purplefox**
### Description

Fixes #3513

This is the first PR in a stack.

This PR provides the implementation of multiple table functions.

### Testing done
New QTT tests as appropriate

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksql-execution/src/main/java/io/confluent/ksql/execution/function/udtf/KudtfFlatMapper.java [hunk 1]
```diff
@@ -28,20 +29,32 @@
 @Immutable
 public class KudtfFlatMapper implements ValueMapper<GenericRow, Iterable<GenericRow>> {
```
Comment (agavra) [inline ksql-execution/src/main/java/io/confluent/ksql/execution/function/udtf/KudtfFlatMapper.java]: Do we have a unit test for this class? I'd expect some more detailed unit tests around this behavior than just QTT

File: ksql-execution/src/main/java/io/confluent/ksql/execution/function/udtf/KudtfFlatMapper.java [hunk 2]
```diff
@@ -28,20 +29,32 @@
...
+    final List<Iterator<Object>> iters = new ArrayList<>(tableFunctionAppliers.size());
+    int maxLength = 0;
+    for (TableFunctionApplier applier: tableFunctionAppliers) {
+      final List<Object> exploded = applier.apply(row);
+      iters.add(exploded.iterator());
+      maxLength = Math.max(maxLength, exploded.size());
+    }
+    final List<GenericRow> rows = new ArrayList<>(maxLength);
+    for (int i = 0; i < maxLength; i++) {
+      final List<Object> newRow = new ArrayList<>(row.getColumns());
```
Comment (agavra) [inline ksql-execution/src/main/java/io/confluent/ksql/execution/function/udtf/KudtfFlatMapper.java]: perhaps these are some premature optimizations, but this is done per-row and I don't think it takes a readability hit so it's probably better to:
1. avoid using the copy-constructor of `ArrayList` (it actually calls toArray() on the source list, which does a fully array copy, then calls another Array.copyOf to copy the data over)
2. pre-size the array appropriately since we know the length ahead of time (`new ArrayList<>(row.getColumns().size() + iters.length())`)

File: ksql-execution/src/main/java/io/confluent/ksql/execution/function/udtf/KudtfFlatMapper.java [hunk 3]
```diff
@@ -28,20 +29,32 @@
...
-      rows.add(new GenericRow(arrayList));
+    final List<Iterator<Object>> iters = new ArrayList<>(tableFunctionAppliers.size());
+    int maxLength = 0;
+    for (TableFunctionApplier applier: tableFunctionAppliers) {
+      final List<Object> exploded = applier.apply(row);
+      iters.add(exploded.iterator());
+      maxLength = Math.max(maxLength, exploded.size());
+    }
+    final List<GenericRow> rows = new ArrayList<>(maxLength);
+    for (int i = 0; i < maxLength; i++) {
```
Comment (agavra) [inline ksql-execution/src/main/java/io/confluent/ksql/execution/function/udtf/KudtfFlatMapper.java]: might be good to have an inline comment describing the zipjoin behavior of multiple UDTFs to set expectations reading this code (or point to some documentation about how multiple table functions work)

File: ksql-execution/src/main/java/io/confluent/ksql/execution/function/udtf/KudtfFlatMapper.java (hunk 2 above)
Comment (agavra) [reply ksql-execution/src/main/java/io/confluent/ksql/execution/function/udtf/KudtfFlatMapper.java]: Ah I had misread the code - there's a possible extra array copy here, but I guess it's unlikely:
```java
            // c.toArray might (incorrectly) not return Object[] (see 6260652)
            if (elementData.getClass() != Object[].class)
                elementData = Arrays.copyOf(elementData, size, Object[].class);
```
But you are probably right, this is unlikely. We could also call `ensureCapacity` before adding the values from the iter to avoid resizing the array more than once.

Comment (purplefox) [reply ksql-execution/src/main/java/io/confluent/ksql/execution/function/udtf/KudtfFlatMapper.java]: > Ah I had misread the code - there's a possible extra array copy here:
> 
> ```java
>             // c.toArray might (incorrectly) not return Object[] (see 6260652)
>             if (elementData.getClass() != Object[].class)
>                 elementData = Arrays.copyOf(elementData, size, Object[].class);
> ```
> 
> But you are probably right, this is unlikely. We could also call `ensureCapacity` before adding the values from the iter to avoid resizing the array more than once.
> 
> EDIT: looks like that bug was fixed in jdk8, so we should be safe agains the extra array copy.

I actually debugged this and this line of code doesn't get triggered in the case of an ArrayList to ArrayList copy :)

But yes, I would be very surprised if we ever get to the point in the project where find this is causing a bottleneck. Most likely other things on the execution path take orders of magnitude more time. 

With respect to resizing. This is very unlikely to happen as table functions are added at the end unless the user has a very large number of table functions! And even if a resize did occur the cost of the allocation will most likely be a pointer addition on the TLAB so almost negligible ;) 

TLDR, we shouldn't worry about it, we have _way_ bigger fish to fry! ;)

File: ksql-execution/src/main/java/io/confluent/ksql/execution/function/udtf/KudtfFlatMapper.java (hunk 1 above)
Comment (purplefox) [reply ksql-execution/src/main/java/io/confluent/ksql/execution/function/udtf/KudtfFlatMapper.java]: Added

File: ksql-execution/src/main/java/io/confluent/ksql/execution/function/udtf/KudtfFlatMapper.java (hunk 2 above)
Comment (agavra) [reply ksql-execution/src/main/java/io/confluent/ksql/execution/function/udtf/KudtfFlatMapper.java]: hehe fair enough! I thought my suggestion was harmless at first, but it grew to take a readability hit so I'm with you.
